DDP communication hook: skip dividing grads by world_size if hook registered.

sinannasir opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42400 DDP communication hook: skip dividing grads by world_size if hook registered.**

@mcarilli spotted that in the original DDP communication hook design described in [39272](https://github.com/pytorch/pytorch/issues/39272), the hooks receive grads that are already predivided by world size.

It makes sense to skip the divide completely if hook registered. The hook is meant for the user to completely override DDP communication. For example, if the user would like to implement something like GossipGrad, always dividing by the world_size would not be a good idea.

We also included a warning in the register_comm_hook API as:
> GradBucket bucket's tensors will not be predivided by world_size. User is responsible to divide by the world_size in case of operations like allreduce.

**Update:** We discovered and fixed a bug with the sparse tensors case. See new unit test called `test_ddp_comm_hook_sparse_gradients` and changes in `reducer.cpp`.

Differential Revision: [D22883905](https://our.internmc.facebook.com/intern/diff/D22883905/)

sinannasir committed:
DDP communication hook: skip dividing grads by world_size if hook registered.

@mcarilli spotted that in the original DDP communication hook design described in [39272](https://github.com/pytorch/pytorch/issues/39272), the hooks receive grads that are already predivided by world size.

It makes sense to skip the divide completely if hook registered. The hook is meant for the user to completely override DDP communication. For example, if the user would like to implement something like GossipGrad, always dividing by the world_size would not be a good idea.

We also included a warning in the register_comm_hook API as:
> GradBucket bucket's tensors will not be predivided by world_size. User is responsible to divide by the world_size in case of operations like allreduce.

Differential Revision: [D22883905](https://our.internmc.facebook.com/intern/diff/D22883905/)

[ghstack-poisoned]

pritamdamania87 commented on torch/nn/parallel/distributed.py:
```diff
@@ -639,6 +639,10 @@ def _register_comm_hook(self, state: object, hook: callable):
                              c10d.ProcessGroupNCCL.work.
 
+        .. warning ::
+            GradBucket bucket's tensors will not be predivided by world_size. User is
```
nit: mentioning bucket twice here, probably say GradBucket's tensors will ....

sinannasir replied:
just replaced with grad bucket to be consistent with other parts of the API.

pritamdamania87 commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -385,7 +391,10 @@ void Reducer::mark_variable_ready_sparse(VariableIndex index) {
     replica.contents = grad;
-    replica.contents.div_(process_group_->getSize());
+    // See Note [DDP Communication Hook]
+    if (comm_hook_ == nullptr) {
+      replica.contents.div_(process_group_->getSize());
+    }
```
We should probably test the communication hook with sparse tensors as well. Using [nn.EmbeddingBag](https://pytorch.org/docs/stable/generated/torch.nn.EmbeddingBag.html) with `sparse=True` will generate sparse gradients for you.

sinannasir replied:
@pritamdamania87 Thanks for this comment! It was quite useful to discover and fix a bug with sparse tensors case. 
I think we should just copy in case of sparse gradients, since `bucket_view` is not used. Please see `test_ddp_comm_hook_sparse_gradients` and changes in `reducer.cpp` to make it work.

pritamdamania87 reviewed (changes requested):
Looks good overall, requesting changes since I just realized that we don't test the communication hook with sparse tensors.

sinannasir committed:
Update on "DDP communication hook: skip dividing grads by world_size if hook registered."


@mcarilli spotted that in the original DDP communication hook design described in [39272](https://github.com/pytorch/pytorch/issues/39272), the hooks receive grads that are already predivided by world size.

It makes sense to skip the divide completely if hook registered. The hook is meant for the user to completely override DDP communication. For example, if the user would like to implement something like GossipGrad, always dividing by the world_size would not be a good idea.

We also included a warning in the register_comm_hook API as:
> GradBucket bucket's tensors will not be predivided by world_size. User is responsible to divide by the world_size in case of operations like allreduce.

Differential Revision: [D22883905](https://our.internmc.facebook.com/intern/diff/D22883905/)

[ghstack-poisoned]

sinannasir committed:
Update on "DDP communication hook: skip dividing grads by world_size if hook registered."


@mcarilli spotted that in the original DDP communication hook design described in [39272](https://github.com/pytorch/pytorch/issues/39272), the hooks receive grads that are already predivided by world size.

It makes sense to skip the divide completely if hook registered. The hook is meant for the user to completely override DDP communication. For example, if the user would like to implement something like GossipGrad, always dividing by the world_size would not be a good idea.

We also included a warning in the register_comm_hook API as:
> GradBucket bucket's tensors will not be predivided by world_size. User is responsible to divide by the world_size in case of operations like allreduce.

**Update:** We discovered and fixed a bug with the sparse tensors case. See new unit test called `test_ddp_comm_hook_sparse_gradients` and changes in `reducer.cpp`.

Differential Revision: [D22883905](https://our.internmc.facebook.com/intern/diff/D22883905/)

[ghstack-poisoned]

sinannasir committed:
Update on "DDP communication hook: skip dividing grads by world_size if hook registered."


@mcarilli spotted that in the original DDP communication hook design described in [39272](https://github.com/pytorch/pytorch/issues/39272), the hooks receive grads that are already predivided by world size.

It makes sense to skip the divide completely if hook registered. The hook is meant for the user to completely override DDP communication. For example, if the user would like to implement something like GossipGrad, always dividing by the world_size would not be a good idea.

We also included a warning in the register_comm_hook API as:
> GradBucket bucket's tensors will not be predivided by world_size. User is responsible to divide by the world_size in case of operations like allreduce.

**Update:** We discovered and fixed a bug with the sparse tensors case. See new unit test called `test_ddp_comm_hook_sparse_gradients` and changes in `reducer.cpp`.

Differential Revision: [D22883905](https://our.internmc.facebook.com/intern/diff/D22883905/)

[ghstack-poisoned]

sinannasir committed:
Update on "DDP communication hook: skip dividing grads by world_size if hook registered."


@mcarilli spotted that in the original DDP communication hook design described in [39272](https://github.com/pytorch/pytorch/issues/39272), the hooks receive grads that are already predivided by world size.

It makes sense to skip the divide completely if hook registered. The hook is meant for the user to completely override DDP communication. For example, if the user would like to implement something like GossipGrad, always dividing by the world_size would not be a good idea.

We also included a warning in the register_comm_hook API as:
> GradBucket bucket's tensors will not be predivided by world_size. User is responsible to divide by the world_size in case of operations like allreduce.

**Update:** We discovered and fixed a bug with the sparse tensors case. See new unit test called `test_ddp_comm_hook_sparse_gradients` and changes in `reducer.cpp`.

Differential Revision: [D22883905](https://our.internmc.facebook.com/intern/diff/D22883905/)

[ghstack-poisoned]
